### PR TITLE
ticlient: clear store address info when receiving StoreNotMatch 

### PR DESCRIPTION
**Pull request by AndreMouche**
https://github.com/pingcap/tidb/issues/2294 @disksing @coocood

### CODE REVIEW

**Comment by coocood** [comment]
What problem does this PR solve?

**Comment by AndreMouche** [comment]
to fix [issue_2294](https://github.com/pingcap/tidb/issues/2294) @coocood

File: server/server.go [hunk 1]
```diff
@@ -201,7 +201,7 @@ func (s *Server) onConn(c net.Conn) {
 	}
 
 	s.rwlock.Lock()
-	s.clients[conn.connectionID] = conn
+	s.clients[conn.connectionID] = conn // TODO:why unreuse?
```
Comment (disksing) [inline server/server.go]: This line of code change is not related to this PR, you may open an issue for it.

File: store/tikv/mock-tikv/cluster.go [hunk 2]
```diff
@@ -112,6 +112,13 @@ func (c *Cluster) RemoveStore(storeID uint64) {
 	delete(c.stores, storeID)
 }
 
+// UpdateStoreAddr update store address for cluster
```
Comment (disksing) [inline store/tikv/mock-tikv/cluster.go]: UpdateStoreAddr **updates** store address for cluster.

File: store/tikv/region_cache.go [hunk 3]
```diff
@@ -57,6 +57,14 @@ type RPCContext struct {
 	Addr   string
 }
 
+// GetStoreID returns StoreID
```
Comment (disksing) [inline store/tikv/region_cache.go]: Please add a `.` at the end of comments.

File: store/tikv/region_cache.go [hunk 4]
```diff
@@ -254,7 +262,11 @@ func (c *RegionCache) GetStoreAddr(bo *Backoffer, id uint64) (string, error) {
 		return store.Addr, nil
 	}
 	c.storeMu.RUnlock()
+	return c.ReloadStoreAddr(bo, id)
+}
 
+// ReloadStoreAddr reload and save store's address
+func (c *RegionCache) ReloadStoreAddr(bo *Backoffer, id uint64) (string, error) {
```
Comment (disksing) [inline store/tikv/region_cache.go]: Why extract this new method?

File: store/tikv/region_cache.go [hunk 5]
```diff
@@ -269,6 +281,13 @@ func (c *RegionCache) GetStoreAddr(bo *Backoffer, id uint64) (string, error) {
 	return addr, nil
 }
 
+// ClearStoreByID clear store from cache with storeID
```
Comment (disksing) [inline store/tikv/region_cache.go]: clears

File: store/tikv/region_cache_test.go [hunk 6]
```diff
@@ -253,3 +253,24 @@ func (s *testRegionCacheSuite) TestRequestFail(c *C) {
...
+	testKey := []byte("test_key")
+	testValue := []byte("test_value")
+	err := client.Put(testKey, testValue)
+	c.Assert(err, IsNil)
+	// tikv-server reports `StoreNotMatch` And retry
+	store1Addr := s.storeAddr(s.store1)
+	s.cluster.UpdateStoreAddr(s.store1, s.storeAddr(s.store2))
+	s.cluster.UpdateStoreAddr(s.store2, store1Addr)
+
+	expectValue, err := client.Get(testKey)
```
Comment (disksing) [inline store/tikv/region_cache_test.go]: It's not expect value but the actual value that load from server. I suppose name it `val` is enough.

File: server/server.go (hunk 1 above)
Comment (AndreMouche) [inline server/server.go]: ok

File: store/tikv/region_cache.go (hunk 4 above)
Comment (AndreMouche) [inline store/tikv/region_cache.go]: Need to use this method alone one day ?  Just like it may need reload store address forcibly in the feature. @disksing

**Comment by disksing** [comment]
LGTM
